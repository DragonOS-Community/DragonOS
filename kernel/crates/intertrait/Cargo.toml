[package]
name = "intertrait"
version = "0.2.2"
authors = ["CodeChain Team <hi@codechain.io>"]
license = "GPLv2(for code modified by dragonos) MIT OR Apache-2.0"
description = "Allow for inter-trait casting"
edition = "2018"
repository = "https://github.com/CodeChain-io/intertrait"
documentation = "https://docs.rs/intertrait"
readme = "README.md"
categories = ["rust-patterns"]
keywords = ["trait", "cast", "any"]
include = ["src/**/*", "Cargo.toml", "LICENSE-*", "README.md"]


[dependencies]
linkme = "=0.3.27"
hashbrown = "0.13.2"
intertrait-macros = { version = "=0.2.2", path = "macros" }

[target.'cfg(not(target_os = "none"))'.dependencies]
once_cell = "1.4"

[dev-dependencies]
trybuild = "1.0"
doc-comment = "0.3"

