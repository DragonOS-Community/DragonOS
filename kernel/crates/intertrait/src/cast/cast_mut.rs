use crate::{caster, CastFrom};

/// A trait that is blanket-implemented for traits extending `CastFrom` to allow for casting
/// of a trait object for it behind an mutable reference to a trait object for another trait
/// implemented by the underlying value.
///
/// # Examples
/// ```
/// # use intertrait::*;
/// use intertrait::cast::*;
///
/// # #[cast_to(Greet)]
/// # struct Data;
/// # trait Source: CastFrom {}
/// # trait Greet {
/// #     fn greet(&self);
/// # }
/// # impl Greet for Data {
/// #    fn greet(&self) {
/// #        println!("Hello");
/// #    }
/// # }
/// impl Source for Data {}
/// let mut data = Data;
/// let source: &mut dyn Source = &mut data;
/// let greet = source.cast::<dyn Greet>();
/// greet.unwrap().greet();
/// ```
pub trait CastMut {
    /// Casts a mutable reference to this trait into that of type `T`.
    fn cast<T: ?Sized + 'static>(&mut self) -> Option<&mut T>;
}

/// A blanket implementation of `CastMut` for traits extending `CastFrom`.
impl<S: ?Sized + CastFrom> CastMut for S {
    fn cast<T: ?Sized + 'static>(&mut self) -> Option<&mut T> {
        let any = self.mut_any();
        let caster = caster::<T>((*any).type_id())?;
        (caster.cast_mut)(any).into()
    }
}
