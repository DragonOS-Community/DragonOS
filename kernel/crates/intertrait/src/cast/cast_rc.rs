use crate::{caster, CastFrom};
use alloc::rc::Rc;

/// A trait that is blanket-implemented for traits extending `CastFrom` to allow for casting
/// of a trait object for it behind an `Rc` to a trait object for another trait
/// implemented by the underlying value.
///
/// # Examples
/// ```
/// # use std::rc::Rc;
/// # use intertrait::*;
/// use intertrait::cast::*;
///
/// # #[cast_to(Greet)]
/// # struct Data;
/// # trait Source: CastFrom {}
/// # trait Greet {
/// #     fn greet(&self);
/// # }
/// # impl Greet for Data {
/// #    fn greet(&self) {
/// #        println!("Hello");
/// #    }
/// # }
/// impl Source for Data {}
/// let data = Data;
/// let source = Rc::new(data);
/// let greet = source.cast::<dyn Greet>();
/// greet.unwrap_or_else(|_| panic!("must not happen")).greet();
/// ```
pub trait CastRc {
    /// Casts an `Rc` for this trait into that for type `T`.
    fn cast<T: ?Sized + 'static>(self: Rc<Self>) -> Result<Rc<T>, Rc<Self>>;
}

/// A blanket implementation of `CastRc` for traits extending `CastFrom`.
impl<S: ?Sized + CastFrom> CastRc for S {
    fn cast<T: ?Sized + 'static>(self: Rc<Self>) -> Result<Rc<T>, Rc<Self>> {
        match caster::<T>((*self).type_id()) {
            Some(caster) => Ok((caster.cast_rc)(self.rc_any())),
            None => Err(self),
        }
    }
}
