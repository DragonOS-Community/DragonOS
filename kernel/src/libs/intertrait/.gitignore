# Cargo lock in subs
**/Cargo.lock

# Generated by Cargo
**/target/

**/*.rs.bk
**/*.iml
.idea/
.vscode/
/db/
/snapshot/
/log/
/keys/
/test/log/

# macOS
.DS_store
