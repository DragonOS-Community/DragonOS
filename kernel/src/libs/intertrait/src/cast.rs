//! `cast` module contains traits to provide `cast` method for various references
//! and smart pointers.
//!
//! In source files requiring casts, import all of the traits as follows:
//!
//! ```ignore
//! use intertrait::cast::*;
//! ```
//!
//! Since there exists single trait for each receiver type, the same `cast` method is overloaded.
mod cast_arc;
mod cast_box;
mod cast_mut;
mod cast_rc;
mod cast_ref;

pub use cast_arc::*;
pub use cast_box::*;
pub use cast_mut::*;
pub use cast_rc::*;
pub use cast_ref::*;
