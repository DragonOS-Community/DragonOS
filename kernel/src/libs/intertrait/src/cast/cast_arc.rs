use alloc::sync::Arc;

use crate::{caster, CastFromSync};

/// A trait that is blanket-implemented for traits extending `CastFrom` to allow for casting
/// of a trait object for it behind an `Rc` to a trait object for another trait
/// implemented by the underlying value.
///
/// # Examples
/// ```
/// # use std::sync::Arc;
/// # use intertrait::*;
/// use intertrait::cast::*;
///
/// # #[cast_to([sync] Greet)]
/// # struct Data;
/// # trait Source: CastFrom {}
/// # trait Greet {
/// #     fn greet(&self);
/// # }
/// # impl Greet for Data {
/// #    fn greet(&self) {
/// #        println!("Hello");
/// #    }
/// # }
/// impl Source for Data {}
/// let data = Data;
/// let source = Arc::new(data);
/// let greet = source.cast::<dyn Greet>();
/// greet.unwrap_or_else(|_| panic!("must not happen")).greet();
/// ```
pub trait CastArc {
    /// Casts an `Arc` for this trait into that for type `T`.
    fn cast<T: ?Sized + 'static>(self: Arc<Self>) -> Result<Arc<T>, Arc<Self>>;
}

/// A blanket implementation of `CastArc` for traits extending `CastFrom`, `Sync`, and `Send`.
impl<S: ?Sized + CastFromSync> CastArc for S {
    fn cast<T: ?Sized + 'static>(self: Arc<Self>) -> Result<Arc<T>, Arc<Self>> {
        match caster::<T>((*self).type_id()) {
            Some(caster) => Ok((caster.cast_arc)(self.arc_any())),
            None => Err(self),
        }
    }
}
