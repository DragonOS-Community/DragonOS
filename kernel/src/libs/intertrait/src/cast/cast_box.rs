use alloc::boxed::Box;

use crate::{caster, CastFrom};

/// A trait that is blanket-implemented for traits extending `CastFrom` to allow for casting
/// of a trait object for it behind a `Box` to a trait object for another trait
/// implemented by the underlying value.
///
/// # Examples
/// ```
/// # use intertrait::*;
/// use intertrait::cast::*;
///
/// # #[cast_to(Greet)]
/// # struct Data;
/// # trait Source: CastFrom {}
/// # trait Greet {
/// #     fn greet(&self);
/// # }
/// # impl Greet for Data {
/// #    fn greet(&self) {
/// #        println!("Hello");
/// #    }
/// # }
/// impl Source for Data {}
/// let data = Box::new(Data);
/// let source: Box<dyn Source> = data;
/// let greet = source.cast::<dyn Greet>();
/// greet.unwrap_or_else(|_| panic!("casting failed")).greet();
/// ```
pub trait CastBox {
    /// Casts a box to this trait into that of type `T`. If fails, returns the receiver.
    fn cast<T: ?Sized + 'static>(self: Box<Self>) -> Result<Box<T>, Box<Self>>;
}

/// A blanket implementation of `CastBox` for traits extending `CastFrom`.
impl<S: ?Sized + CastFrom> CastBox for S {
    fn cast<T: ?Sized + 'static>(self: Box<Self>) -> Result<Box<T>, Box<Self>> {
        match caster::<T>((*self).type_id()) {
            Some(caster) => Ok((caster.cast_box)(self.box_any())),
            None => Err(self),
        }
    }
}
