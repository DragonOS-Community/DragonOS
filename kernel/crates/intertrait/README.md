# Intertrait

We forked this lib from [intertrait](https://github.com/CodeChain-io/intertrait/) (revision d5d6dcb), and modified it to support `no_std` environment.

## Notice

The modified version is licensed under GPLv2 and later, while the original version is licensed under MIT/Apache license.(Codes modified by us are licensed under GPLv2 and later.)


This library provides direct casting among trait objects implemented by a type.

In Rust, a trait object for a sub-trait of [`std::any::Any`] can be downcast to a concrete type at runtime
if the type is known. But no direct casting between two trait objects (i.e. without involving the concrete type
of the backing value) is possible (even no coercion from a trait object for a trait to that for its super-trait yet).

With this crate, any trait object for a sub-trait of [`CastFrom`] can be cast directly to a trait object
for another trait implemented by the underlying type if the target traits are registered beforehand
with the macros provided by this crate.

# Dependencies
Add the following two dependencies to your `Cargo.toml`:

```toml
[dependencies]
intertrait = "0.2"
linkme = "=0.3.27"
```

The `linkme` dependency is required due to the use of `linkme` macro in the output of `intertrait` macros.

# Usage

```rust
use intertrait::*;
use intertrait::cast::*;

struct Data;

trait Source: CastFrom {}

trait Greet {
    fn greet(&self);
}

#[cast_to]
impl Greet for Data {
    fn greet(&self) {
        println!("Hello");
    }
}

impl Source for Data {}

fn main() {
    let data = Data;
    let source: &dyn Source = &data;
    let greet = source.cast::<dyn Greet>();
    greet.unwrap().greet();
}
```

Target traits must be explicitly designated beforehand. There are three ways of doing it:

### `#[cast_to]` to `impl` item
The trait implemented is designated as a target trait.

```rust
use intertrait::*;

struct Data;
trait Greet { fn greet(&self); }

#[cast_to]
impl Greet for Data {
    fn greet(&self) {
        println!("Hello");
    }
}
```

### `#[cast_to(Trait)]` to type definition
For the type, the traits specified as arguments to the `#[cast_to(...)]` attribute are designated as target traits.

```rust
use intertrait::*;

trait Greet { fn greet(&self); }

impl Greet for Data {
    fn greet(&self) {
        println!("Hello");
    }
}

#[cast_to(Greet, std::fmt::Debug)]
#[derive(std::fmt::Debug)]
struct Data;
```

### `castable_to!(Type => Trait1, Trait2)`
For the type, the traits following `:` are designated as target traits.

```rust
use intertrait::*;

#[derive(std::fmt::Debug)]
struct Data;
trait Greet { fn greet(&self); }
impl Greet for Data {
    fn greet(&self) {
        println!("Hello");
    }
}
// Only in an item position due to the current limitation in the stable Rust.
// https://github.com/rust-lang/rust/pull/68717
castable_to!(Data => Greet, std::fmt::Debug);

fn main() {}
```

## `Arc` Support
`std::sync::Arc` is unique in that it implements `downcast` method only on `dyn Any + Send + Sync + 'static'.
To use with `Arc`, the following steps should be taken:

* Mark source traits with [`CastFromSync`] instead of [`CastFrom`]
* Add `[sync]` flag to `#[cast_to]` and `castable_to!` as follows:
  ```ignore
  #[cast_to([sync])]
  #[cast_to([sync] Trait1, Trait2)]
  castable_to!(Type => [sync] Trait, Trait2);
  ```

# How it works
First of all, [`CastFrom`] trait makes it possible to retrieve an object of [`std::any::Any`]
from an object for a sub-trait of [`CastFrom`]. 

And the macros provided by `intertrait` generates trampoline functions for downcasting a trait object
for [`std::any::Any`] back to its concrete type and then creating a trait object for the target trait from it.

Those trampoline functions are aggregated into a global registry
using [`linkme`](https://github.com/dtolnay/linkme/) crate, which involves no (generally discouraged)
life-before-main trick. The registry is keyed with a pair of [`TypeId`]s, which are those of the concrete type
backing a trait object for a sub-trait of [`CastFrom`] and the target trait (the actual implementation
is a bit different here, but conceptually so).

In the course, it doesn't rely on any unstable Rust implementation details such as the layout of trait objects
that may be changed in the future.

# Credits
`intertrait` has taken much of its core ideas from the great [`traitcast`](https://github.com/bch29/traitcast) crate.

# License
The modified version is licensed under GPLv2 and later, while the original version is licensed under MIT/Apache license.(Codes modified by us are licensed under GPLv2 and later.)

Modified version(revision 0.2.0):
 * GPLv2 and later (You can find the full text of the license in the root directory of this repository.)

Original version(revision d5d6dcb):
 * Apache License, Version 2.0
   ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license
   ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

