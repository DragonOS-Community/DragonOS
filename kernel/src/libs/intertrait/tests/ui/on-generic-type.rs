use intertrait::*;
use intertrait::cast::*;
use std::marker::PhantomData;

#[cast_to(Greet)]
struct Data<T: 'static> {
    phantom: PhantomData<T>,
}

trait Source: CastFrom {}

trait Greet {
    fn greet(&self);
}

impl<T: 'static> Greet for Data<T> {
    fn greet(&self) {
        println!("Hello");
    }
}

impl<T: 'static> Source for Data<T> {}

fn main() {
    let data = Data::<i32> {
        phantom: PhantomData,
    };
    let source: &dyn Source = &data;
    let greet = source.cast::<dyn Greet>();
    greet.unwrap().greet();
}
