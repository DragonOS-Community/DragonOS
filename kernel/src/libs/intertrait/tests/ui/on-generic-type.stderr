error: #[cast_to(..)] can't be used on a generic type definition
 --> tests/ui/on-generic-type.rs:6:12
  |
6 | struct Data<T: 'static> {
  |            ^^^^^^^^^^^^
